src/main/java/org/reactome/addlinks/db/ReferenceObjectCache.java

```diff
@@ -106,183 +106,217 @@ private static synchronized void buildReferenceCaches(String className, Map<Stri
 					}
 				}
-				catch (InvalidAttributeException e)
+				catch (ArrayIndexOutOfBoundsException e)
 				{
-					logger.error("Could not get the \"{}\" attribute for {} because it not valid for this object.", ReactomeJavaConstants.referenceDatabase, referenceObject);
-				}
-				catch (Exception e)
-				{
-					logger.error("Could not get the \"{}\" attribute for {}. Reason: {}", ReactomeJavaConstants.identifier, referenceObject, e.getMessage());
+					logger.error("ArrayIndexOutOfBounds was caught! Identifier that triggered this was: {}; className: {}; objectCacheByIdentifier has {} items; objectCacheByIdentifier[identifier] has {} items", identifier, className, objectCacheByIdentifier.size(), objectCacheByIdentifier.containsKey(identifier) ? objectCacheByIdentifier.get(identifier).size() : " *THAT IDENTIFIER IS NOT IN THAT MAP!* ");
 					e.printStackTrace();
+					throw new Error(e);
 				}
+
+			}
+		}
+		catch (InvalidAttributeException e)
+		{
+			logger.error("Object {} does not have an identifier attribute, error: {}", referenceObject, e.getMessage());
+			e.printStackTrace();
+		}
+		catch (Exception e)
+		{
+			e.printStackTrace();
+			throw new Error(e);
+		}
+	}
+
+	/**
+	 * Adds an object to a cache that is indexed by ReferenceDatabase.
+	 * @param cacheByRefDB - the cache to add to.
+	 * @param referenceObject - the object to add.
+	 */
+	private static void addObjectToRefDBCache(Map<String, List<GKInstance>> cacheByRefDB, GKInstance referenceObject)
+	{
+		String refDBID;
+		try
+		{
+			refDBID = String.valueOf(((GKInstance) referenceObject.getAttributeValue(ReactomeJavaConstants.referenceDatabase)).getDBID());
+			if (! cacheByRefDB.containsKey(refDBID))
+			{
+				List<GKInstance> byRefDb = Collections.synchronizedList( new LinkedList<GKInstance>() );
+				byRefDb.add(referenceObject);
+				cacheByRefDB.put(refDBID, byRefDb);
+			}
+			else
+			{
+				cacheByRefDB.get(refDBID).add(referenceObject);
+			}
+		}
+		catch (InvalidAttributeException e)
+		{
+			logger.error("Could not get the \"{}\" attribute for {} because it not valid for this object.", ReactomeJavaConstants.referenceDatabase, referenceObject);
+		}
+		catch (Exception e)
+		{
+			logger.error("Could not get the \"{}\" attribute for {}. Reason: {}", ReactomeJavaConstants.identifier, referenceObject, e.getMessage());
+			e.printStackTrace();
+		}
+	}
+
+	/**
+	 * Adds an object to a cache that is indexed by species.
+	 * @param className - the className of the object.
+	 * @param cacheBySpecies - the cache to add the object to.
+	 * @param referenceObject - the object to add.
+	 */
+	private static void addObjectToSpeciesCache(String className, Map<String, List<GKInstance>> cacheBySpecies, GKInstance referenceObject)
+	{
+		// Sets do not allow duplicates.
+		Set<String> allSpecies = null;
+		//ReferenceRNASequence objects don't always have Species info, for some reason, so we need to get that from the associated ReferenceGeneProduct
```

Comment:
This might be worth bringing up to the curators on a Tuesday call to make sure it's not a data error.  If all ReferenceRNASequences should have species, there should probably be a QA check.

Reply:
I think I spoke to someone (maybe you?) waaaaaay back when I first wrote this and it seemed like an allowable condition, at the time. You can raise it, if you want.

Reply:
I can think of cases where it may be okay, but it's probably best to confirm with the curators.